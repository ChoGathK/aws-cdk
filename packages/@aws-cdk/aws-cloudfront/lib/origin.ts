import { IBucket } from '@aws-cdk/aws-s3';
import { Token } from '@aws-cdk/core';
import { CfnDistribution } from './cloudfront.generated';

/**
 * Represents a CloudFront Origin and its behaviors.
 */
export interface IOrigin {
  /**
   * A unique identifier for the origin. This value must be unique within the distribution.
   */
  readonly id: string;

  /**
   * The domain name for the origin.
   */
  readonly domainName: string;

  /**
   * Creates and returns the CloudFormation representation of this origin.
   */
  renderOrigin(): CfnDistribution.OriginProperty;
}

/**
 * Properties to be used to create an Origin. Prefer to use one of the Origin.from* factory methods rather than
 * instantiating an Origin directly from these properties.
 */
export interface OriginProps {
  /**
   * A unique identifier for the origin. This value must be unique within the distribution.
   *
   * @default - Assigned automatically.
   */
  readonly id?: string;

  /**
   * The domain name of the Amazon S3 bucket or HTTP server origin.
   */
  readonly domainName: string;
}

/**
 * Represents a distribution origin, that describes the Amazon S3 bucket, HTTP server (for example, a web server),
 * Amazon MediaStore, or other server from which CloudFront gets your files.
 */
export class Origin implements IOrigin {

  /**
   * Creates a pre-configured origin for a S3 bucket.
   * If this bucket has been configured for static website hosting, then `fromWebsiteBucket` should be used instead.
   *
   * An Origin Access Identity will be created and granted read access to the bucket, unless **TODO**.
   *
   * @param bucket the bucket to act as an origin.
   */
  public static fromBucket(bucket: IBucket): Origin {
    return new Origin({ domainName: bucket.bucketRegionalDomainName, id: bucket.bucketName });
  }

  /**
   * Creates a pre-configured origin for a S3 bucket, where the bucket has been configured for website hosting.
   *
   * An Origin Access Identity will be created and granted read access to the bucket, unless **TODO**.
   *
   * @param bucket the bucket to act as an origin.
   */
  public static fromWebsiteBucket(bucket: IBucket): Origin {
    return new Origin({ domainName: bucket.bucketWebsiteDomainName, id: bucket.bucketName });
  }

  public readonly domainName: string;
  public readonly id: string;

  constructor(props: OriginProps) {
    this.domainName = props.domainName;
    this.id = props.id || Token.asString(undefined);
  }

  public renderOrigin(): CfnDistribution.OriginProperty {
    return {
      domainName: this.domainName,
      id: this.id,
      s3OriginConfig: {},
    };
  }

}