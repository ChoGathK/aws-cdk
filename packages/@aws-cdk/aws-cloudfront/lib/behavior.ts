import { CfnDistribution } from './cloudfront.generated';
import { Origin } from './origin';
import { ViewerProtocolPolicy } from './web_distribution';

/**
 * The HTTP methods that the Behavior will accept requests on.
 */
export class AllowedMethods {
  /** HEAD and GET */
  public static readonly ALLOW_GET_HEAD = new AllowedMethods(['GET' ,'HEAD']);
  /** HEAD, GET, and OPTIONS */
  public static readonly ALLOW_GET_HEAD_OPTIONS = new AllowedMethods(['GET', 'HEAD', 'OPTIONS']);
  /** All supported HTTP methods */
  public static readonly ALLOW_ALL = new AllowedMethods(['GET', 'HEAD', 'OPTIONS', 'PUT', 'PATCH', 'POST', 'DELETE']);

  /** HTTP methods supported */
  public readonly methods: string[];

  private constructor(methods: string[]) { this.methods = methods; }
}

/**
 * Options for adding a behavior to an Origin.
 */
export interface CacheBehaviorOptions {
  /**
   * HTTP methods to allow for this behavior.
   *
   * @default GET and HEAD
   */
  readonly allowedMethods?: AllowedMethods;
}

/**
 * Properties for specifying custom behaviors for origins.
 */
export interface CacheBehaviorProps extends CacheBehaviorOptions {
  /**
   * The pattern (e.g., `images/*.jpg`) that specifies which requests to apply the behavior to.
   * There must be exactly one behavior associated with each `Distribution` that has a path pattern
   * of '*', which acts as the catch-all default behavior.
   */
  readonly pathPattern: string;

  /**
   * The origin that you want CloudFront to route requests to when they match this cache behavior.
   */
  readonly origin: Origin;
}

/**
 * Allows configuring a variety of CloudFront functionality for a given URL path pattern.
 *
 * Note: This really should simply by called 'Behavior', but this name is already taken by the legacy
 * CloudFrontWebDistribution implementation.
 */
export class CacheBehavior {

  /**
   * The pattern (e.g., `images/*.jpg`) that specifies which requests to apply the behavior to.
   */
  public readonly pathPattern: string;
  private readonly origin: Origin;

  constructor(private readonly props: CacheBehaviorProps) {
    this.origin = props.origin;
    this.pathPattern = props.pathPattern;

    this.origin._attachBehavior(this);
  }

  /**
   * Creates and returns the CloudFormation representation of this behavior.
   * This renders as a "CacheBehaviorProperty" regardless of if this is a default
   * cache behavior or not, as the two are identical except that the pathPattern
   * is omitted for the default cache behavior.
   *
   * @internal
   */
  public _renderBehavior(): CfnDistribution.CacheBehaviorProperty {
    return {
      targetOriginId: this.origin.id,
      pathPattern: this.pathPattern,
      allowedMethods: this.props.allowedMethods?.methods ?? undefined,
      forwardedValues: { queryString: false },
      viewerProtocolPolicy: ViewerProtocolPolicy.ALLOW_ALL,
    };
  }

}