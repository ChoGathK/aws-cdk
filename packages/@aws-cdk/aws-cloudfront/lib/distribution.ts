import * as acm from '@aws-cdk/aws-certificatemanager';
import * as s3 from '@aws-cdk/aws-s3';
import { Construct, IResource, Lazy, Resource, Stack, Token } from '@aws-cdk/core';
import { CfnDistribution } from './cloudfront.generated';
import { IOrigin, Origin } from './origin';
import { ViewerProtocolPolicy } from './web_distribution';

/**
 * Interface for CloudFront distributions
 */
export interface IDistribution extends IResource {
  /**
   * The domain name of the Distribution, such as d111111abcdef8.cloudfront.net.
   *
   * @attribute
   */
  readonly domainName: string;

  /**
   * The distribution ID for this distribution.
   *
   * @attribute
   */
  readonly distributionId: string;
}

// TODO - Do we need a BaseDistribution?

/**
 * Properties for a Distribution
 */
export interface DistributionProps {
  /**
   * The primary origin for the distribution.
   */
  readonly origin: IOrigin;

  /**
   * A certificate to associate with the distribution. The certificate must be located in N. Virginia (us-east-1).
   *
   * @default the CloudFront wildcard certificate (*.cloudfront.net) will be used.
   */
  readonly certificate?: acm.ICertificate;
}

/**
 * Attributes used to import a Distribution.
 */
export interface DistributionAttributes {
  /**
   * The generated domain name of the Distribution, such as d111111abcdef8.cloudfront.net.
   *
   * @attribute
   */
  readonly domainName: string;

  /**
   * The distribution ID for this distribution.
   *
   * @attribute
   */
  readonly distributionId: string;
}

/**
 * A CloudFront distribution with associated origin(s) and caching behavior(s).
 */
export class Distribution extends Resource implements IDistribution {

  /**
   * Creates a Distribution construct that represents an external (imported) distribution.
   */
  public static fromDistributionAttributes(scope: Construct, id: string, attrs: DistributionAttributes): IDistribution {
    return new class extends Resource implements IDistribution {
      public readonly domainName: string;
      public readonly distributionId: string;

      constructor() {
        super(scope, id);
        this.domainName = attrs.domainName;
        this.distributionId = attrs.distributionId;
      }
    }();
  }

  /**
   * Creates a Distribution for an S3 Bucket, where the bucket has not been configured for website hosting.
   *
   * This creates a single-origin distribution with a single default behavior.
   */
  public static forBucket(scope: Construct, id: string, bucket: s3.IBucket): Distribution {
    return new Distribution(scope, id, {
      origin: Origin.fromBucket(scope, 'SingleOriginBucket', bucket),
    });
  }

  /**
   * Creates a Distribution for an S3 bucket, where the bucket has been configured for website hosting.
   *
   * This creates a single-origin distribution with a single default behavior.
   */
  public static forWebsiteBucket(scope: Construct, id: string, bucket: s3.IBucket): Distribution {
    return new Distribution(scope, id, {
      origin: Origin.fromWebsiteBucket(scope, 'SingleOriginWebsiteBucket', bucket),
    });
  }

  public readonly domainName: string;
  public readonly distributionId: string;

  /**
   * Default origin of the distribution.
   */
  public readonly origin: IOrigin;
  /**
   * Certificate associated with the distribution, if any.
   */
  public readonly certificate?: acm.ICertificate;

  constructor(scope: Construct, id: string, props: DistributionProps) {
    super(scope, id);

    if (props.certificate) {
      const certificateRegion = Stack.of(this).parseArn(props.certificate.certificateArn).region;
      if (!Token.isUnresolved(certificateRegion) && certificateRegion !== 'us-east-1') {
        throw new Error('Distribution certificates must be in the us-east-1 region.');
      }
    }

    this.origin = props.origin;
    this.certificate = props.certificate;

    const distribution = new CfnDistribution(this, 'CFDistribution', { distributionConfig: {
      enabled: true,
      origins: Lazy.anyValue({ produce: () => this.renderOrigins() }),
      defaultCacheBehavior: {
        forwardedValues: { queryString: false },
        viewerProtocolPolicy: ViewerProtocolPolicy.ALLOW_ALL,
        targetOriginId: this.origin.id,
      },
      viewerCertificate: this.renderViewerCertificate(),
    } });

    this.domainName = distribution.attrDomainName;
    this.distributionId = distribution.ref;
  }

  private renderOrigins(): CfnDistribution.OriginProperty[] {
    return [this.origin.renderOrigin()];
  }

  private renderViewerCertificate(): CfnDistribution.ViewerCertificateProperty | undefined {
    return this.certificate ? { acmCertificateArn: this.certificate.certificateArn } : undefined;
  }

}
