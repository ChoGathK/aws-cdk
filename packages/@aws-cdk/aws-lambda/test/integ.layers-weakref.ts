/// !cdk-integ *

import * as cdk from '@aws-cdk/core';
import * as path from 'path';
import * as lambda from '../lib';

const app = new cdk.App();

const layerStack = new cdk.Stack(app, 'integ-layers-weakref-layerstack');

const layer = new lambda.LayerVersion(layerStack, 'Layer', {
  code: lambda.Code.fromAsset(path.join(__dirname, 'layer-code')),
});

const functionStack = new cdk.Stack(app, 'integ-layers-weakref-functionstack');

new lambda.Function(functionStack, 'Function', {
  code: new lambda.InlineCode('foo'),
  handler: 'index.handler',
  runtime: lambda.Runtime.NODEJS_12_X,
  layers: [ layer ],
});

app.synth();
