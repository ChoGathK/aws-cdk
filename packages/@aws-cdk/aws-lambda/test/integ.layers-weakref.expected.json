[
  {
    "Resources": {
      "LayerB20D2F06": {
        "Type": "AWS::Lambda::LayerVersion",
        "Properties": {
          "Content": {
            "S3Bucket": {
              "Ref": "AssetParameterscff0e7519a19b9d2f9e46c677af6798646b9f5d23b6b47b7ca37890e15a73bbeS3Bucket1A231EBF"
            },
            "S3Key": {
              "Fn::Join": [
                "",
                [
                  {
                    "Fn::Select": [
                      0,
                      {
                        "Fn::Split": [
                          "||",
                          {
                            "Ref": "AssetParameterscff0e7519a19b9d2f9e46c677af6798646b9f5d23b6b47b7ca37890e15a73bbeS3VersionKeyFC7F06B6"
                          }
                        ]
                      }
                    ]
                  },
                  {
                    "Fn::Select": [
                      1,
                      {
                        "Fn::Split": [
                          "||",
                          {
                            "Ref": "AssetParameterscff0e7519a19b9d2f9e46c677af6798646b9f5d23b6b47b7ca37890e15a73bbeS3VersionKeyFC7F06B6"
                          }
                        ]
                      }
                    ]
                  }
                ]
              ]
            }
          }
        }
      },
      "SSMExportRefLayerB20D2F06E5F12C7F": {
        "Type": "AWS::SSM::Parameter",
        "Properties": {
          "Name": "/stacks/integ-layers-weakref-layerstack/RefLayerB20D2F06",
          "Description": "[cdk] exported from stack \"integ-layers-weakref-layerstack\" for use as parameter in a different stack",
          "Type": "String",
          "Value": {
            "Ref": "LayerB20D2F06"
          }
        }
      }
    },
    "Parameters": {
      "AssetParameterscff0e7519a19b9d2f9e46c677af6798646b9f5d23b6b47b7ca37890e15a73bbeS3Bucket1A231EBF": {
        "Type": "String",
        "Description": "S3 bucket for asset \"cff0e7519a19b9d2f9e46c677af6798646b9f5d23b6b47b7ca37890e15a73bbe\""
      },
      "AssetParameterscff0e7519a19b9d2f9e46c677af6798646b9f5d23b6b47b7ca37890e15a73bbeS3VersionKeyFC7F06B6": {
        "Type": "String",
        "Description": "S3 key for asset version \"cff0e7519a19b9d2f9e46c677af6798646b9f5d23b6b47b7ca37890e15a73bbe\""
      },
      "AssetParameterscff0e7519a19b9d2f9e46c677af6798646b9f5d23b6b47b7ca37890e15a73bbeArtifactHashFFD7BA29": {
        "Type": "String",
        "Description": "Artifact hash for asset \"cff0e7519a19b9d2f9e46c677af6798646b9f5d23b6b47b7ca37890e15a73bbe\""
      }
    }
  },
  {
    "Resources": {
      "FunctionServiceRole675BB04A": {
        "Type": "AWS::IAM::Role",
        "Properties": {
          "AssumeRolePolicyDocument": {
            "Statement": [
              {
                "Action": "sts:AssumeRole",
                "Effect": "Allow",
                "Principal": {
                  "Service": "lambda.amazonaws.com"
                }
              }
            ],
            "Version": "2012-10-17"
          },
          "ManagedPolicyArns": [
            {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    "Ref": "AWS::Partition"
                  },
                  ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
                ]
              ]
            }
          ]
        }
      },
      "Function76856677": {
        "Type": "AWS::Lambda::Function",
        "Properties": {
          "Code": {
            "ZipFile": "foo"
          },
          "Handler": "index.handler",
          "Role": {
            "Fn::GetAtt": [
              "FunctionServiceRole675BB04A",
              "Arn"
            ]
          },
          "Runtime": "nodejs12.x",
          "Layers": [
            {
              "Ref": "SsmParameterValuestacksinteglayersweakreflayerstackRefLayerB20D2F06"
            }
          ]
        },
        "DependsOn": [
          "FunctionServiceRole675BB04A"
        ]
      }
    },
    "Parameters": {
      "SsmParameterValuestacksinteglayersweakreflayerstackRefLayerB20D2F06": {
        "Type": "AWS::SSM::Parameter::Value<String>",
        "Default": "/stacks/integ-layers-weakref-layerstack/RefLayerB20D2F06"
      }
    }
  }
]